### PR TITLE
Stepper: Apply extended props to `calypso_signup_step_start`

### PR DESCRIPTION
**Pull request by chriskmnds**
Part of addressing https://github.com/Automattic/wp-calypso/issues/94108
Branched off https://github.com/Automattic/wp-calypso/pull/93941

## Proposed Changes

* Adds `STEPPER_TRACKS_EVENT_SIGNUP_STEP_START`
* Implements props extension for `calypso_signup_step_start` event
* Some additional refactors to `useStepRouteTracking`. Removes `recordStepStart` method - this was a minimal abstraction that's only relevant to `useStepRouteTracking` (attempting to even reuse this outside will probably cause tracking inconsistencies)

DO NOT MERGE - depended on ongoing discussions in https://github.com/Automattic/wp-calypso/issues/94175 (similar/same problem)

## Why are these changes being made?

* Allow flows to extend the default props tracked by Stepper for all the major/framework events

## Testing Instructions

* test `/setup/onboarding`
  * ensure tracking of `calypso_signup_step_start` when walking through steps in Stepper works like before
  * should fire once per step render (that's not caused by a change in design, intent, or variant)

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx [hunk 1]
```diff
@@ -25,27 +30,44 @@ const useIsRequestingSelectedSite = ( siteSlugOrId: string | number, site: SiteD
...
+				intent,
+				is_in_hosting_flow: isAnyHostingFlow( flow.name ),
+				...( design && { assembler_source: getAssemblerSource( design ) } ),
+				...( flow.variantSlug && { flow_variant: flow.variantSlug } ),
+				...signupStepStartProps,
+			} );
+		},
+
+		// We leave out intent and design from the dependency list, due to the ONBOARD_STORE being reset in the exit flow.
+		// The store reset causes these values to become empty, and may trigger this event again.
```
Comment (alshakero) [inline client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx]: Hopefully, we'll move design, intent, and `is_in_hosting_flow` out of here into the flows.

File: client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx [hunk 2]
```diff
@@ -25,27 +30,44 @@ const useIsRequestingSelectedSite = ( siteSlugOrId: string | number, site: SiteD
...
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+		[ flow.name, flow.variantSlug, signupStepStartProps ]
```
Comment (alshakero) [inline client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx]: Is there a risk of `signupStepStartProps` causing too many run of the effect?

**Review by alshakero — COMMENTED** [review]
Looks great. Left one question.

File: client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx (hunk 1 above)
Comment (chriskmnds) [reply client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx]: Also hopefully by the end of these refactors we'll have a generalised utility that will record these values by default (once all the framework events are handled similarly)

File: client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx (hunk 2 above)
Comment (chriskmnds) [reply client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx]: Good point. Yes, there is that risk. So the way we have things right now, and once the [steps param is cleaned up](https://github.com/Automattic/wp-calypso/issues/93808#issue-2480245630), this value should be memoized as long as the flow memoizes the initial input. Unfortunately the latter is required - and unless we can come up with some other way, we should make it clear in a doc somewhere (to always memo the return).

I'll give it more thought, but that's the same risk with wherever these extended props are used (they are an object).

Comment (chriskmnds) [reply client/landing/stepper/declarative-flow/internals/components/step-route/hooks/use-step-route-tracking/index.tsx]: Ok. I've thought this over a bit. I'm not sure I have any great/proper solution in mind, at least not something that won't require deeper refactors.

`useTracksEventProps` being a hook, querying and owning state, any changes from there will propagate to here (memo or no memo, the problem is the same - by memoizing we eliminate the rerenders with same state, but state changes will propagate in any other scenario). The same situation exists in other places right now, such as [useSignupStartTracking](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/declarative-flow/internals/hooks/use-sign-up-start-tracking/index.tsx).

I have some ideas for a deeper fix, but for the moment, the best I can think of is, either/or:

1. Exclude `signupStepStartProps` from the dependencies array. This would lead to inconsistencies in what gets logged, as only the first/current version of these props will be exposed from the callback. It ultimately makes the hook behave in a near-static fashion. Same with `variant` and `design` here that we also explicitly ignore. 
2. Leave as-is and expect that multiple events will be recorded until state settles.

On the first point - it is technically the same as what also happens in [useFlowAnalytics](https://github.com/Automattic/wp-calypso/blob/trunk/client/landing/stepper/declarative-flow/internals/hooks/use-flow-analytics/index.tsx), which can also lead to incomplete/non-finalized data - it's just done differently there (we control it via the session storage, and things like `variant` or `search` params are treated as nearly static). and we will run into same crossroad once we introduce `useExtendedProps` there (I have a PR in the works for that).

@alshakero I believe you are AFK now, so cc'ing @Automattic/dotcom-stepper ?

**Comment by travisw** [comment]
Hey @chriskmnds 👋 
I wanted to take a look at this out of curiosity. Being mostly unfamiliar with the code and rusty in React, I want to see if I understand the changes in the PR and approaches you laid out above. (no guarantee that any feedback I give here will be helpful😃)

It sounds like the purpose of the PR is to allow individual flows to add extra Tracks properties to the `calypso_signup_step_start` event, which will get added to the event in addition to the more "global" properties defined here. This helps keep things more modular and minimizes the bloat that gets added to this file. Is that right?

For the approaches...

> 1. Exclude signupStepStartProps from the dependencies array. This would lead to inconsistencies in what gets logged, as only the first/current version of these props will be exposed from the callback. It ultimately makes the hook behave in a near-static fashion. Same with variant and design here that we also explicitly ignore.

How I interpret this is: because `signupStepStartProps` can change over the life of the component, if we don't include it as a dependency of the effect, then these changes won't each trigger the effect to be executed. The upside is that we'd only fire the event once. The downside is that we could potentially have incorrect or incomplete properties being included in the event (I think this would be limited to the extra properties that flows pass in and wouldn't affect the global ones defined in this file?)

> 2. Leave as-is and expect that multiple events will be recorded until state settles.

This sounds like if we leave it as-is, again because `signupStepStartProps` can change, then being a dependency of the effect would cause it to trigger the effect to get exectuted on each of those changes. The upside is that we'd have correct or complete properties in at least one of the fired events. The down side is that we'd potentially end up firing the event multiple times.

Is my understanding close?

If so, then one (maybe naive) question is: why would `signupStepStartProps` have the potential to change values, leading to this issue?

> This would lead to inconsistencies in what gets logged, as only the first/current version of these props will be exposed from the callback

In other words, why would later versions of these props exist beyond that first/current one? I'm guessing something potentially gets called multiple times with different values?

**Comment by travisw** [comment]
If my understanding of the purpose of the PR is correct, then another question that comes to mind is: **do we definitely want or need this functionality?**

From a data perspective, making it too easy to append extra properties to existing and important events could lead to confusing or muddy data. Before adding more properties to an existing event, it might be worth asking whether the properties apply broadly enough to warrant their inclusion.

I think there's an argument to be made for not making it easy for properties of (some) events to easily be appended to, and instead try to enforce a bit more control and consistency over the data. Depending on the property, maybe there are already other ways to measure the same thing or maybe it's just a simple convenience like `is_in_hosting_flow` that should maybe be questioned anyway. I've seen events where basic properties like user_id or blog_id have been added on top of the built-in versions of the same thing that already exist. That type of thing should probably be avoided.

On the other hand I get the need for flexible code. And given that teams don't always talk with one another or fully understand a particular architecture, any inflexible code ecan easily turn into ugly and bloated code.

I'm curious what others think.

**Comment by chriskmnds** [comment]
Hey @travisw . Happy to address all your questions. I'll try and be concise.

> It sounds like the purpose of the PR is to allow individual flows to add extra Tracks properties to the calypso_signup_step_start event, which will get added to the event in addition to the more "global" properties defined here. This helps keep things more modular and minimizes the bloat that gets added to this file. Is that right?

> If my understanding of the purpose of the PR is correct, then another question that comes to mind is: **do we definitely want or need this functionality?**

I had the same question initially (whether we need this or not). I'm fairly convinced that we do need it. Let me explain. It's all to do with separating and decoupling concerns. 

- First, there is no difference technically (at a lower level) whether the flow defines these props in its definition/implementation or if they are added to these corresponding framework hooks. The origin of the data is the same and part of the same life cycle - as well as the point of consumption (triggering the event). So the rest are just layers/abstractions on top.
- The idea is to decentralize the logic, so that we don't end up with flow implementations messing with framework-level hooks and files, more so in ambiguous locations. 
  - Flows now have a localized and central place to define what they need to add/remove/update from what gets recorded (and uniformly across all the events). We do this already for other events, and this PR just adds the above event to the mix (nothing more really). A lot of the props for the various events may naturally pertain to things that are specific to some flow. 
  - The alternative can be messy: either the flow creating a separate event and finding a way to record it at the same time, or the framework hooks containing flow context/conditions - both of which can get unwieldy. 
  - There is also a performance penalty if we choose to record all the props for all the flows (e.g. a flow may require some async operation, like querying an endpoint for metadata) - we don't want that to happen for every flow. So some conditioning/separation is vital.

> How I interpret this is: because signupStepStartProps can change over the life of the component, if we don't include it as a dependency of the effect, then these changes won't each trigger the effect to be executed. The upside is that we'd only fire the event once. The downside is that we could potentially have incorrect or incomplete properties being included in the event 

Correct. But it's more general, not specific to these "extended props".

> (I think this would be limited to the extra properties that flows pass in and wouldn't affect the global ones defined in this file?)

Not exactly. Technically, this is the same for anything that we record and originates from some state variable/prop (whether it's a hook/component's local state, some wp-data store, or a React Query hook). Unless something is defined statically (e.g. not state) then it is a variable thing in the scope of the app. State starts with a default value and may be subject to arbitrary changes (whatever the app reacts to).

I think the above answers the remaining of your comments from https://github.com/Automattic/wp-calypso/pull/93983#issuecomment-2332827717, right? 🤔 

----------

One thing I wanted to highlight from my thoughts in the other issue (see https://github.com/Automattic/wp-calypso/issues/94175#issuecomment-2331219520):

We cannot change the fact that some state property may change throughout the life cycle of a flow or step. I think that's too arbitrary to attempt to. What we can do - IMO - is to either/or (I'm sure more options exist though):

1. Adopt a system over some predefined boundary (i.e. a step's life cycle) and capture all state changes within. The idea of a queuing system with timestamped events comes to mind here. This wouldn't be easy. See my comment for why.
2. Adopt a different mental model over what we capture. Consider whether we do in fact want to capture anything more than what we see **at the exact moment of triggering the event**.

The latter I think may be crucial and will actually lead to simplest solution (to just ignore the dependency). This was my reasoning:

> > as well as whether we do indeed want the latest state to be recorded always
>
> I think this is also critical, but may be "overthinking". For example, site may be recorded as undefined at the beginning of step-1 (before the site is created), and get a concrete value at the end of it (after it was created) - and looks fine to my eyes, because "site does not exist at beginning of step-1" (and we'd not want to record it or retrigger the event with it).

**Comment by MicBosi** [comment]
Hello @chriskmnds 👋 

I have a newbie Stepper question: is it correct to say that Stepper flow/step implementations can already customise the props for all the events listed [here](https://github.com/Automattic/wp-calypso/pull/93983/files#diff-3e4dd91dc201fd303883eb947545b02dc1ac55679af86e0cc0e9260a8de95c1eR23), besides the `calypso_signup_step_start` event added in this PR?

**Comment by travisw** [comment]
I appreciate all the detail @chriskmnds - very helpful.

> State starts with a default value and may be subject to arbitrary changes (whatever the app reacts to).

> We cannot change the fact that some state property may change throughout the life cycle of a flow or step.

> Adopt a different mental model over what we capture. Consider whether we do in fact want to capture anything more than what we see *at the exact moment of triggering the event*.

> as well as whether we do indeed want the latest state to be recorded always

> For example, site may be recorded as undefined at the beginning of step-1 (before the site is created), and get a concrete value at the end of it (after it was created)

Maybe it would be helpful to better understand some specific scenarios to know whether there's any meaningful risk, or if just using the value that exists when the event is triggered, like you said, is enough.

For the site example that you gave, do you have a scenario in mind where this would happen? Like exactly which step in which flow? Are there other steps and properties where you foresee this same tradeoff. I'm wondering what's an example of an initial value and a final value. Maybe it's such a corner case that it's not a huge concern.

**Review by escapemanuele — APPROVED** [review]
Tested it and it all works correctly 👍

**Comment by chriskmnds** [comment]
Let's merge. The dependency is currently ignored, so anything async in these props will not be recorded. I will follow up with a PR to ignore `calypso_signup_start` extended props also (as those can cause a rerender/re-record when supplied). Reference issue/discussion: https://github.com/Automattic/wp-calypso/issues/95106#issuecomment-2391818351

Lots of confusion around this area. Just sometimes, the hardest parts are plainly non-technical 😬
